### PR TITLE
Add collision checking for footprint without using subscibers.

### LINKED ISSUE #1666: Class which calculates footprint cost
## Feature request
Implement a class which is feeds on COSTMAP2D and provides functionality such as calculates cost of footprint at particular position.
#### Implementation considerations

* I think it would be similar to these class from ROS Navigation
https://github.com/ros-planning/navigation/blob/melodic-devel/base_local_planner/src/costmap_model.cpp

### PR DESCRIPTION
**Pull request by shivaang12**
---
Closes #1666 
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1666 |
| Primary OS tested on |  |
| Robotic platform tested on |  |

---

## Description of contribution in a few bullet points
* Add collision checking feature without subscribers
* Stripped `CollisionChecker` class to support this feature. 

## Description of documentation updates required from your changes

---

## Future work that may be required in bullet points
* Maybe adding more rigorous unit tests.
* Integration tests

### CODE REVIEW

File: nav2_costmap_2d/CMakeLists.txt [hunk 1]
```diff
@@ -105,6 +106,19 @@ target_link_libraries(nav2_costmap_2d_client
   nav2_costmap_2d_core
 )
 
+add_library(nav2_costmap_2d_standalone SHARED
```
Comment (SteveMacenski) [inline nav2_costmap_2d/CMakeLists.txt:109]: This name must change, its not descriptive of what's in it

File: nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2020 Shivang Patel
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp:1]: Please keep the names of people that had derived work included here too. I'd actually argue for this case since its mostly a refactor, your name should not be included in this paticular set of files.

File: nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
+#include "tf2/utils.h"
+#pragma GCC diagnostic pop
+
+namespace nav2_costmap_2d
+{
+typedef std::vector<geometry_msgs::msg::Point> Footprint;
+
+class CollisionBase
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp:39]: I think this should be called `FootprintCollisionChecker` and we should rename the collision checker using subscriptions as something else. This is the "core" collision checking capability and that's just s ROS wrapper for it.

File: nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+#include "nav2_costmap_2d/footprint_subscriber.hpp"
+#include "nav2_util/robot_utils.hpp"
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
+#include "tf2/utils.h"
+#pragma GCC diagnostic pop
+
+namespace nav2_costmap_2d
+{
+class CollisionFootprint : public CollisionBase
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp:38]: Since this uses the collision tools, shouldn't all those be in the core collision checker? The external API of gtting the footprint cost seems to me to be a requirement for that.

Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp:38]: I'd suggest putting all this in the collision base class.

File: nav2_costmap_2d/include/nav2_costmap_2d/collision_checker.hpp [hunk 5]
```diff
@@ -34,9 +36,8 @@
 
 namespace nav2_costmap_2d
 {
-typedef std::vector<geometry_msgs::msg::Point> Footprint;
 
-class CollisionChecker
+class CollisionChecker : public CollisionBase
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/collision_checker.hpp:40]: Why does this inherit? Wouldn't it be more reasonable for the this subscription-based collision checker to use the collision checker on the given costmap from the subscription? 

In C++ convention is always to compose rather than inherit when possible.

File: nav2_costmap_2d/test/unit/collision_footprint_test.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  p4.x = 0.0;
+  p4.y = -0.5;
+
+  nav2_costmap_2d::Footprint footprint = {p1, p2, p3, p4};
+
+  nav2_costmap_2d::CollisionFootprint collision_footprint(costmap_);
+
+  auto value = collision_footprint.footprintCostWithPose(5.0, 5.0, 0.0, footprint);
+
+  EXPECT_NEAR(value, 0.0, 0.001);
```
Comment (SteveMacenski) [inline nav2_costmap_2d/test/unit/collision_footprint_test.cpp:32]: Shouldn't this work with 0.0 and not near? This is deterministic

File: nav2_costmap_2d/test/unit/collision_footprint_test.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
+#include <string>
+#include <vector>
+#include <memory>
+
+#include "gtest/gtest.h"
+#include "nav2_costmap_2d/collision_footprint.hpp"
+
+TEST(collision_footprint, footprintCostWithPose)
```
Comment (SteveMacenski) [inline nav2_costmap_2d/test/unit/collision_footprint_test.cpp:8]: These tests should more accurately test that things are working, having an empty costmap is going to always work. Can you add some obstacles around the robot but just barely not touching to make it try to strain things if anything is really wrong?

Comment (SteveMacenski) [reply nav2_costmap_2d/test/unit/collision_footprint_test.cpp:8]: A test with obstacles all around it but not touching it, move the footprint over 1 pixel, and check that its now in collision.

File: nav2_costmap_2d/src/collision_footprint.cpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+{
+  return costmap_->getCost(x, y);
+}
+
+void CollisionFootprint::worldToMap(double wx, double wy, unsigned int & mx, unsigned int & my)
+{
+  costmap_->worldToMap(wx, wy, mx, my);
+}
+
+double CollisionFootprint::footprintCostWithPose(
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/collision_footprint.cpp:46]: Maybe `footprintCostAtPose`? `With` reads a little awkwardly.

File: nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp (hunk 2 above)
Comment (shivaang12) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp:1]: Sorry about that. I was actually going to add `modified by` section at the end.

File: nav2_costmap_2d/CMakeLists.txt (hunk 1 above)
Comment (shivaang12) [reply nav2_costmap_2d/CMakeLists.txt:109]: I have very poor naming skills

File: nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp (hunk 4 above)
Comment (shivaang12) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp:38]: I main reason it that original `CollisionChecker` classes uses exceptions. I don't want that in footprint checker

File: nav2_costmap_2d/test/unit/collision_footprint_test.cpp (hunk 6 above)
Comment (shivaang12) [reply nav2_costmap_2d/test/unit/collision_footprint_test.cpp:32]: I can do that but I have this OCD with comparing floating types. So, as a measure I use near.

File: nav2_costmap_2d/src/collision_footprint.cpp (hunk 8 above)
Comment (shivaang12) [reply nav2_costmap_2d/src/collision_footprint.cpp:46]: I told you, I am not so good with naming things ;)

File: nav2_costmap_2d/test/unit/collision_footprint_test.cpp (hunk 6 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/test/unit/collision_footprint_test.cpp:32]: Sure - we can keep it

File: nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp (hunk 4 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp:38]: Changing the inheritance and making some methods virtual overrides breaks ABI anyhow, if you're going to break it; break it

Comment (shivaang12) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp:38]: Don't get me wrong here, but this design doesn't break `CollisionChecker`'s API (well at least I believe, and also I haven't modified its tests and they are still working and passing without any sweat).

If I have permission to break then let me see where its going to affect. Because the main thing I am worried about is - exceptions. I am worried that somewhere some code might rely on these exception will fail immediately. I think I should have investigated beforehand but that's a mistake on my part.

Anyways, let me see where `CollisionChecker` is used and then I will make the changes accordingly.

Comment (shivaang12) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp:38]: I think this is very strange. No where in our stack this is being used. Even DWB has similar implementation of this in form of `obstacle_footprint` critic. I think we should remove this and make it purely non-subscription based.

File: nav2_costmap_2d/CMakeLists.txt (hunk 1 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/CMakeLists.txt:109]: `nav2_costmap_2d_collision_checker` or something

File: nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp (hunk 4 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp:38]: The subscriptions are used by the recoveries for their collision detection

File: nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp (hunk 3 above)
Comment (shivaang12) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp:39]: You want to call it `CollisionCheckerSub`?

Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp:39]: Call what? I think this comment is for the core collision stuff. If you're talking about the subscription based one in another file, I think `ExternalCollisionChecker` perhaps? I'm also not great with naming. Something that gets it across its for use outside of the costmaps where you need to subscribe to a costmap topic

Comment (shivaang12) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp:39]: I am fine with `ExternalCollisionChecker`!

Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/collision_base.hpp:39]: @shivaang12 `CostmapTopicCollisionChecker` lets just go full on wordy.

**Comment by shivaang12** [comment]
@SteveMacenski Can you please take a look to this PR? Thanks!

**Comment by SteveMacenski** [comment]
@shivaang12 you have a conflict to resolve

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_topic_collision_checker.hpp [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2019 Intel Corporation
+// Copyright (c) 2019 Intel Corporationl
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_topic_collision_checker.hpp:1]: 1?

File: nav2_costmap_2d/src/costmap_topic_collision_checker.cpp [hunk 10]
```diff
@@ -107,68 +99,7 @@ Footprint CollisionChecker::getFootprint(const geometry_msgs::msg::Pose2D & pose
...
-  } else if (cost == NO_INFORMATION) {
-    RCLCPP_DEBUG(rclcpp::get_logger(name_), "Map Cell: [%d, %d]", x, y);
-    throw IllegalPoseException(name_, "Footprint Hits Unknown Region.");
-  }
-
-  return cost;
-}
-
-void CollisionChecker::unorientFootprint(
+void CostmapTopicCollisionChecker::unorientFootprint(
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_topic_collision_checker.cpp:102]: What's this unorient footprint for?

File: nav2_costmap_2d/test/unit/footprint_collision_checker_test.cpp [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    std::make_shared<nav2_costmap_2d::Costmap2D>(100, 100, 0.1, 0, 0, 0);
+
+  nav2_costmap_2d::FootprintCollisionChecker collision_checker(costmap_);
+
+  auto value = collision_checker.pointCost(50, 50);
+
+  EXPECT_NEAR(value, 0.0, 0.001);
+}
+
+TEST(collision_footprint, worldToMap)
```
Comment (SteveMacenski) [inline nav2_costmap_2d/test/unit/footprint_collision_checker_test.cpp:61]: This doesn't test anything, you set the background to 0 so of course any cell it selects will be 0. You should add some single cell value of non-zero and then check from the world to map that that works.

File: nav2_costmap_2d/test/unit/footprint_collision_checker_test.cpp [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+#include "gtest/gtest.h"
+#include "nav2_costmap_2d/footprint_collision_checker.hpp"
+
+TEST(collision_footprint, footprintCostWithPose)
+{
+  std::shared_ptr<nav2_costmap_2d::Costmap2D> costmap_ =
+    std::make_shared<nav2_costmap_2d::Costmap2D>(100, 100, 0.1, 0, 0, 0);
+
+  geometry_msgs::msg::Point p1;
```
Comment (SteveMacenski) [inline nav2_costmap_2d/test/unit/footprint_collision_checker_test.cpp:27]: This test is all zeros, so of course it will work. Add obstacles around the robot and then try that. My recommendation:
- Add a cost 1 cell to the right in the middle
- Add a cost cell to the left on a corner
- Test that in this pose its OK
- Test that when you move the footprint 1 cell to the left its in collision (checks point cost)
- Test that when you move the footprint 1 cell to the right its in collision (checks line cost)

File: nav2_costmap_2d/test/unit/footprint_collision_checker_test.cpp [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+  nav2_costmap_2d::FootprintCollisionChecker collision_checker(costmap_);
+
+  auto value = collision_checker.footprintCostAtPose(5.0, 5.0, 0.0, footprint);
+  EXPECT_NEAR(value, 0.0, 0.001);
+
+  auto up_value = collision_checker.footprintCostAtPose(5.0, 4.9, 0.0, footprint);
+  EXPECT_NEAR(up_value, 254.0, 0.001);
+
+  auto down_value = collision_checker.footprintCostAtPose(5.0, 5.2, 0.0, footprint);
+  EXPECT_NEAR(down_value, 254.0, 0.001);
```
Comment (shivaang12) [inline nav2_costmap_2d/test/unit/footprint_collision_checker_test.cpp:145]: @SteveMacenski I am testing by moving the footprint here.

Comment (SteveMacenski) [reply nav2_costmap_2d/test/unit/footprint_collision_checker_test.cpp:145]: This is what I would call an integration test, it makes sure everything plays well together, but since outside that window its all lethal, you cant be sure if the point cost and the line cost individually are working. My comment above is just adding 2 specifically located costs to specifically test the point and line cost.

Comment (SteveMacenski) [reply nav2_costmap_2d/test/unit/footprint_collision_checker_test.cpp:145]: This test could pass even though the line cost would fail.

File: nav2_costmap_2d/src/costmap_topic_collision_checker.cpp (hunk 10 above)
Comment (shivaang12) [reply nav2_costmap_2d/src/costmap_topic_collision_checker.cpp:102]: It is used to get normalized footprint from the positional footprint. You get positional footprint when you subscribe to the footprint topic. This class first normalizes the footprint from the subscription and then pushes it to the obstacle checking mechanism.

One thing is strange, this whole thing is first transform the coordinated footprint to normalized footprint and then transform it back to coordinated footprint using current coordinate of robot.

Let me check if we can eliminate the double transformation thing.

**Comment by shivaang12** [comment]
Checking the system test works without whole transform back and forth thing.

**Comment by shivaang12** [comment]
@SteveMacenski I have added the requested test. Please check the bottom most test. I want this to get this through as soon as possible. Thanks!

File: nav2_costmap_2d/src/costmap_topic_collision_checker.cpp (hunk 10 above)
Comment (shivaang12) [reply nav2_costmap_2d/src/costmap_topic_collision_checker.cpp:102]: I tried removing the `unorientedfootprint` and all the test do pass without it. Please take a look. At first glance this function seem redundant but might be a hidden secret behind its existence. Please review it and let me know.

Comment (SteveMacenski) [reply nav2_costmap_2d/src/costmap_topic_collision_checker.cpp:102]: @shivaang12 it is used in the function above in master, it looks like you commented it out just 10 lines above. I think that's a regression.

**Comment by SteveMacenski** [comment]
@shivaang12 you haven't fixed any of my review comments

**Comment by shivaang12** [comment]
I have reverted the `unorientedFootprint` usage. I have added the setvalue costmap worldtomap test. Also, the point and line test has been added.

**Comment by SteveMacenski** [comment]
@shivaang12 rebase - the test failure is from something that's been fixed.

**Comment by SteveMacenski** [comment]
Woah - what's all that new stuff @shivaang12 ? I just needed you to rebase

**Comment by shivaang12** [comment]
Ahhh!! may be I screwed up something. I got three way merge thing when rebased as someone removed lib export of nav2_costmap_markers. Let me look into it.

**Comment by shivaang12** [comment]
Again rebase to Master.

**Comment by SteveMacenski** [comment]
Ship it
